### ISSUE TITLE
Memtest86+ 7.00 passes, but four other memory tests fail. How come?

### ISSUE THREAD

**Issue by jfikar** [memtest86plus/memtest86plus#371]
Hi,
I have an old MacBookPro 2011 and I tried to upgrade its DDR3 RAM to 16GB. I have run several memory tests and only Memtest86+ passes, which is strange. Is the Memtest86+ somewhat less sensitive to errors?

Test passed:
Memtest86+ 7.00

![memtest86plus](https://github.com/memtest86plus/memtest86plus/assets/9248946/ea2070cc-340d-4cdb-97b5-0d710e9898ce)

Tests failed:
Memtest86 v10.6 free
Memtester 4.6.0 `memtester 15G 10`
Apple build in diagnostics (boot with G pressed).

See attached pictures (sorry for the quality, will take better ones, if needed).
![memtest86](https://github.com/memtest86plus/memtest86plus/assets/9248946/2327f86d-3a94-4970-bbbc-a6dd3b3541e6)
![apple](https://github.com/memtest86plus/memtest86plus/assets/9248946/c0f2fddc-2284-4c34-badc-8b5948dbd11f)
![memtester](https://github.com/memtest86plus/memtest86plus/assets/9248946/591572c9-238b-45d6-a80b-ab5d109e95f2)

If I recall correctly, it had already happened once. On another computer Memtest86+ passes, but Memtest86 fails (these RAM sticks are no more with me, as I returned them).

UPDATE
A minor thing: Memtest86 shows L1 as 64kB, which is IMHO correct, but Memtest86+ shows only 32kB.

**Comment by debrouxl** [comment]
Several notes as I go thinking about this report:
* I've already seen memory errors at random, non-reproducible addresses on computers whose memory access subsystem is running beyond the higher clock speed it can reasonably achieve, because of a motherboard defect, or something like that;
* it's interesting that memtest86+ runs in 4C/8T mode, as indicated by the CPU itself and the ACPI MADT, whereas the crippled no-fee version of Passmark's proprietary, commercial software you're using finds a single CPU... possibly a limitation of that computer's EFI services, since Passmark memtest86 ties itself to EFI, which has both upsides and downsides ?
* I don't know about Apple's diagnostic program, but memtester runs single-threaded as well (and seemingly without setting affinity, in the Debian sid amd64 version I have handy, which should be reasonably current);
* memtest86+ does not attempt to enable turbo mode (SpeedStep, Turbo Boost, etc.), but the 2200 MHz clock reported by Passmark memtest86 may well mean that it doesn't either on that computer ?

=> I'd say that two useful additional data points would be:
* restricting memtest86+ to running on core #0 (the bootstrap processor, BSP), which is probably what Passmark memtest86 does, and Apple's diagnostic program probably does unless it's multi-threaded;
* restricting the memory speed to 1333 MHz: the maximum speed this mobile SNB generation processor supports, anyway (I'd have expected to see IMC working at 1333 MHz, not 1866 MHz ?).
Could you perform such tests as well ? TIA :)

As for 32 KB vs. 64 KB... The i7-2675QM has 32 KB L1D$ + 32 KB L1I$ per core. memtest86+ considers that only the L1D cache is relevant to a program strongly oriented towards data access, because it is technically correct that the tight memory writing and reading loops (the hottest code paths, by far) are much smaller than the L1I$ size. Passmark memtest86 considers that reporting the sum of L1D$ and L1I$ sizes is the way to go. Basically, relevant L1 cache size on the one side, and commercial L1 cache size on the other side :)

**Comment by x86fr** [comment]
That's a very interesting case. 

Testing these memory modules on another motherboard would be very instructive. Do you have any way to do this?

**Comment by jfikar** [comment]
I'm not sure, how to restrict memtest86+ to one core. I've chosen SEQ and it runs very slowly. That's probably expected. And I got two errors so far:

![memtest-fail](https://github.com/memtest86plus/memtest86plus/assets/9248946/03b5eaa9-8ac9-41fb-bfb7-59f8dbc88e40)


Maybe the RR would be faster? As before, it is a random flip of the top bit.

I can't change the memory speed to 1333MHz, there is absolutely no BIOS settings on a Macbook. The only way would be to manipulate SPD on the sticks. The memory was before in Lenovo ThinkPad X220 for years without a problem. So I suppose the RAM is fine. But I don't have the X220 anymore (answers the @x86fr question). And it was running there at 1866MHz. It had similar CPU i5-2520M.

Isn't it strange that single-core mode shows more errors than the multi-core? I would think, it should be the opposite. I had once a problem with an ARM64 board, 4 core CPU. It was failing memtester if I run 4x in parallel. When it I run on a single core, it passed. I even spoke to the memtester author. He said that it is probably dead CPU cache (probably related to the overclocking, which was officially supported though), which gets more stressed during the parallel runs. And that memtester is originally meant to run on a single core.

And yes, the memtester86 says something about EFI limits, when I try to activate more cores.

Thanks for the L1 explanation, it is clear now.

**Comment by debrouxl** [comment]
Yeah, in your screenshots, it's always a flip on the MSB, but the addresses do not seem to reproduce across runs. My gut feeling is that more than in the memory itself, the data corruption occurs elsewhere: during transit, or probably less likely, inside the CPU. Especially if these memory sticks worked in another computer for years, at 1866 MHz, as you write.

Too bad the Macbook variant of a PC isn't more configurable. Two of my regular x86 PC motherboards do not work properly at their maximum speed, but reducing the speed from 400 MHz to 333 MHz and 1333 MHz normal / 1600 MHz OC to 1066 MHz respectively makes them stable under Linux, including Prime95 (mostly benchmark / validation mode, as having old computers perform real work on Mersenne numbers is very inefficient), and usable for testing my memtest86+ developments.

On processors where a single core cannot reach the full memory bandwidth, and for errors which only occur beyond a sufficient memory bandwidth, I do expect multi-core mode to show more errors than single-core mode.
